libsql-server/src/stats.rs

```diff
@@ -52,6 +53,100 @@ impl SlowestQuery {
     }
 }
 
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq, PartialOrd, Ord)]
+pub struct QueryStats {
+    pub elapsed_ms: u64,
+    pub count: u64,
+    pub rows_written: u64,
+    pub rows_read: u64,
+}
+
+impl QueryStats {
+    pub fn new(elapsed_ms: u64, rows_read: u64, rows_written: u64) -> Self {
+        Self {
+            elapsed_ms,
+            count: 1,
+            rows_read,
+            rows_written,
+        }
+    }
+    pub fn empty() -> Self {
+        Self {
+            elapsed_ms: 0,
+            count: 0,
+            rows_read: 0,
+            rows_written: 0,
+        }
+    }
+    pub fn merge(&self, another: &QueryStats) -> Self {
+        Self {
+            elapsed_ms: self.elapsed_ms + another.elapsed_ms,
+            count: self.count + another.count,
+            rows_read: self.rows_read + another.rows_read,
+            rows_written: self.rows_written + another.rows_written,
+        }
+    }
+}
+
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub struct QueriesStats {
+    #[serde(default)]
+    id: Option<Uuid>,
+
+    #[serde(default)]
+    stats_threshold: AtomicU64,
+    #[serde(default)]
+    stats: HashMap<String, QueryStats>,
+}
+
+impl QueriesStats {
+    pub fn new() -> Arc<RwLock<Self>> {
+        let mut this = QueriesStats::default();
+        this.id = Some(Uuid::new_v4());
+        Arc::new(RwLock::new(this))
+    }
+
+    pub fn register_query(&mut self, sql: &String, stat: QueryStats) {
+        let (aggregated, new) = match self.stats.get(sql) {
+            Some(stat) => (stat.clone(), false),
+            None => (QueryStats::empty(), true),
+        };
+
+        let aggregated = aggregated.merge(&stat);
+        if aggregated.elapsed_ms < self.stats_threshold.load(Ordering::Relaxed) {
+            return;
+        }
+
+        self.stats.insert(sql.clone(), aggregated);
+
+        if !new || self.stats.len() <= 30 {
+            return;
+        }
+
+        let mut vec = self.stats.clone().into_iter().collect_vec();
```

Comment:
could we use https://doc.rust-lang.org/stable/std/collections/struct.HashMap.html#method.drain here instead? I think it would be more efficient.

Reply:
My understanding is that drain removes all values from the map. 
I don't want to do this, since I just want to remove a single entry.
I've removed the vec in favor of just iterating and getting the minimum element.